source/extensions/path/uri_template_lib/uri_template_internal.cc

```diff
@@ -0,0 +1,386 @@
+  // Locate the variable pattern to parse.
+  if (pattern.size() < 2 || (pattern)[0] != '{') {
+    return absl::InvalidArgumentError("Invalid variable");
+  }
+  std::vector<absl::string_view> parts = absl::StrSplit(pattern.substr(1), absl::MaxSplits('}', 1));
+  if (parts.size() != 2) {
+    return absl::InvalidArgumentError("Unmatched variable bracket");
+  }
+  absl::string_view unconsumed_pattern = parts[1];
+
+  // Parse the actual variable pattern, starting with the variable name.
+  std::vector<absl::string_view> var_parts = absl::StrSplit(parts[0], absl::MaxSplits('=', 1));
+  if (!isValidVariableName(var_parts[0])) {
+    return absl::InvalidArgumentError("Invalid variable name");
+  }
+  Variable var = Variable(var_parts[0], {});
+
+  // Parse the variable match pattern (if any).
+  if (var_parts.size() < 2) {
+    return ParsedResult<Variable>(var, unconsumed_pattern);
+  }
+  absl::string_view var_patt = var_parts[1];
+  if (var_patt.empty()) {
+    return absl::InvalidArgumentError("Empty variable match");
+  }
+  while (!var_patt.empty()) {
+    absl::variant<Operator, Literal> var_match;
+    if (var_patt[0] == '*') {
+
+      absl::StatusOr<Operator> status = alsoUpdatePattern<Operator>(consumeOperator, &var_patt);
+      if (!status.ok()) {
+        return status.status();
+      }
+      var_match = *std::move(status);
+
+    } else {
+
+      absl::StatusOr<Literal> status = alsoUpdatePattern<Literal>(consumeLiteral, &var_patt);
+      if (!status.ok()) {
+        return status.status();
+      }
+      var_match = *std::move(status);
+    }
+    var.var_match_.push_back(var_match);
+    if (!var_patt.empty()) {
+      if (var_patt[0] != '/' || var_patt.size() == 1) {
+        return absl::InvalidArgumentError("Invalid variable match");
+      }
+      var_patt = var_patt.substr(1);
+    }
+  }
+
+  return ParsedResult<Variable>(var, unconsumed_pattern);
+}
+
+absl::StatusOr<absl::flat_hash_set<absl::string_view>>
+gatherCaptureNames(struct ParsedUrlPattern pattern) {
+  absl::flat_hash_set<absl::string_view> captured_variables;
+
+  for (const ParsedSegment& segment : pattern.parsed_segments_) {
+    if (!absl::holds_alternative<Variable>(segment)) {
+      continue;
+    }
+    if (captured_variables.size() >= kPatternMatchingMaxVariablesPerUrl) {
+      return absl::InvalidArgumentError("Exceeded variable count limit");
+    }
+    absl::string_view var_name = absl::get<Variable>(segment).var_name_;
+
+    if (var_name.size() < kPatternMatchingMinVariableNameLen ||
+        var_name.size() > kPatternMatchingMaxVariableNameLen) {
+      return absl::InvalidArgumentError("Invalid variable length");
+    }
+    if (captured_variables.contains(var_name)) {
+      return absl::InvalidArgumentError("Repeated variable name");
+    }
+    captured_variables.emplace(var_name);
+  }
+
+  return captured_variables;
+}
+
+absl::Status validateNoOperatorAfterTextGlob(struct ParsedUrlPattern pattern) {
```

Comment:
I don't see this in a .h file (or the previous method). Should it be moved into an anon namespace?

nit: should pattern be passed by const&, perhaps?

Reply:
Added to header and used const&